olive/passes/onnx/pipeline/step_utils.py

```diff
@@ -0,0 +1,231 @@
+import importlib as imp
+from typing import Dict, List, Union
+
+import onnx
+from onnxruntime_extensions.tools.pre_post_processing import *  # noqa: F401, F403
+from onnxruntime_extensions.tools.pre_post_processing.utils import create_named_value
+
+from olive.passes.onnx.pipeline import resolve_placeholder
+
+
+def parse_steps(model: onnx.ModelProto, config: List[Dict]):
+    """Parse the config and return a dictionary of step name and its parameters.
+
+    ## Config examples:
+    1. if there io_map is missing, the config could be simplified as:
+        ```
+        [
+            {
+                "ReverseAxis": {
+                    "axis": 2,
+                    "dim_value": 3,
+                    "name": "BGR_to_RGB"
+                }
+            },
+            {
+                "Normalize": {
+                    "normalization_values": [[0.485, 0.229],[0.456,0.224], [0.406,0.225]],
+                    "layout": "CHW"
+                }
+            }
+        ]
+        ```
+
+    2. if there is io_map, the config looks like:
+        ```
+        {
+            "Resize": {
+                "params": {
+                    "resize_to": [
+                        {
+                            "type": "__model_input__",
+                            "input_index": 0,
+                            "dim_index": -1
+                        },
+                        {
+                            "type": "__model_output__",
+                            "output_index": 0,
+                            "dim_index": -2
+                        }
+                    ]
+                },
+                "io_map": [["ConvertImageToBGR",0,0]]
+            }
+        }
+        ```
+    ## Config schema:
+
+    1. simple form:
+         ```
+         { }"step_name": {"param_name": param_value} }
+         ```
+
+    2. full form:
+        ```
+        {
+            "step_name": {
+                "params": {"param_name": param_value},
+                "io_map": [[step_name, input_index, output_index]]
+            }
+        }
+        ```
+
+       The detailed io_map definition and usage is
+       https://github.com/microsoft/onnxruntime-extensions/blob/main/onnxruntime_extensions/tools/Example%20usage%20of%20the%20PrePostProcessor.md#iomapentry-usage
+
+       Please note the io_map is optional, if it is missing, the config is simplified as the simple form.
+
+    3. Customized parameter type for parameter value:
+        ```
+        {
+            "SentencePieceTokenizer": {
+                "tokenizer_param": {
+                    "type": "TokenizerParam",
+                    "params": {
+                        "vocab_or_file": "test_data/sentencepiece/test_model.model"
+                    }
+                }
+            }
+        }
+        ```
+       In this example, the type will be looked in the imported modules. So, please remember to import it before
```

Comment:
As Olive will be run in CLI model. Where to import the corresponding modules by users? in user script?

Reply:
From the time being, Olive doesn't support user customize steps in user_script.py. We can consider add that support in future in case of user has the requirement.

The steps are imported in step_utils.py